### PR TITLE
chore: use $ReadOnlyArray when input array is not modified (fixes #84)

### LINKED ISSUE #84: DataLoader does not allow result as $ReadOnlyArray
I've started getting an error after updating to Flow v0.47:

```
src/factories/loaders/createEventIdByCoordinatesAndMovieIdLoader.js:28
              v----------------------------
 28:   return new DataLoader((queries) => {
 29:     return map(queries, async (query) => {
 30:       return getEventIdsByCoordinatesAndMovieId(
...:
 53:   });
       -^ constructor call
 13: type MapType<T, R> = (values: $ReadOnlyArray<T>, mapper: MapperType<T, R>, configuration?: MapConfigurationType) => Promise<$ReadOnlyArray<R>>;
                                                                                                                                 ^^^^^^^^^^^^^^^^^ read-only array type. This type is incompatible with. See: node_modules/bluefeather/dist/map.js.flow:13
 13: type BatchLoadFn<K, V> = (keys: Array<K>) => Promise<Array<V | Error>>;
                                                          ^^^^^^^^^^^^^^^^ array type. See: node_modules/dataloader/index.js.flow:13

```

`DataLoader` should allow `$ReadOnlyArray` as a valid response. Whats the reason it does not?

### PR DESCRIPTION
**Pull request by gajus**

### CODE REVIEW

**Comment by gajus** [comment]
As far as I can tell, these are the only safe changes.

I have left output type as `Array` as that would be a breaking change/ outside of scope.

I have left comments as `Array` since `$ReadOnlyArray` is a sub-type of `Array`.

cc @wincent

**Comment by gajus** [comment]
The build process is failing because of:

```
/home/travis/build/facebook/dataloader/src/index.js
   13:1   error  Line 13 exceeds the maximum line length of 80  max-len
   13:33  error  '$ReadOnlyArray' is not defined                no-undef
   13:63  error  '$ReadOnlyArray' is not defined                no-undef
  133:18  error  '$ReadOnlyArray' is not defined                no-undef
  309:26  error  '$ReadOnlyArray' is not defined                no-undef

```

Have you considered using https://github.com/gajus/eslint-plugin-flowtype?

I have added `$ReadOnlyArray` to globals as an interim fix.

**Comment by wincent** [comment]
Thanks @gajus. Will merge with a couple of tweaks applied on top.

**Comment by gajus** [comment]
Regarding feee2aa, that was an ESLint error, not FLow.

File: src/index.js [hunk 1]
```diff
@@ -162,7 +162,7 @@ export default class DataLoader<K, V> {
   }
 
   /**
-   * Adds the provided key and value to the cache. If the key already exists, no
```
Comment (gajus) [inline src/index.js]: No idea how this got into the PR. Sorry.

**Comment by wincent** [comment]
> Regarding feee2aa, that was an ESLint error, not Flow.

Bah, should have waited for Travis...

**Comment by gajus** [comment]
@wincent Can this be released please?

I am still getting dozens of:

```
Error: src/factories/createLoaders.js:267
267:     ReservationByReservationUidLoader: new DataLoader(getReservationByReservationUids),
                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ function. This type is incompatible with the expected param type of
 45:     batchLoadFn: BatchLoadFn<K, V>,
                      ^^^^^^^^^^^^^^^^^ function type. See: node_modules/dataloader/index.js.flow:45
  This parameter is incompatible:
    143:     return getByIdsLoader(connection, 'reservation', reservationUids, 'uid');
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type application of identifier `Promise`. Has some incompatible type argument with
     13: type BatchLoadFn<K, V> = (keys: Array<K>) => Promise<Array<V | Error>>;
                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^ type application of identifier `Promise`. See: node_modules/dataloader/index.js.flow:13
      Type argument `R` is incompatible:
         42: ): Promise<$ReadOnlyArray<QueryResultRowType>> => {
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ read-only array type. This type is incompatible with. See: src/utilities/getByIdsLoader.js:42
         13: type BatchLoadFn<K, V> = (keys: Array<K>) => Promise<Array<V | Error>>;
                                                                  ^^^^^^^^^^^^^^^^ array type. See: node_modules/dataloader/index.js.flow:13

```

It is really discouraging to contribute anything if it does not get released for months.

**Comment by gajus** [comment]
@leebyron Any comment here?

**Pull request merged**